File: src/com/fenghuo/notes/db/BackupRestoreUtils.java
package com.fenghuo.notes.db;

import android.content.Context;
import android.os.Environment;

import com.fenghuo.notes.bean.Alarm;
import com.fenghuo.notes.bean.Kind;
import com.fenghuo.notes.bean.MonthAccount;
import com.fenghuo.notes.bean.Note;

import org.json.JSONObject;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.util.HashMap;
import java.util.List;

public class BackupRestoreUtils {

	private Context context;
	private DBAlarmHelper alarmHelper;
	private DBAccountHelper accountHelper;
	private DbKindHelper kindHelper;
	private DBNoteHelper noteHelper;

	private List<Alarm> listalarm;
	private List<Kind> listKind;
	private List<Note> listnote;
	private List<MonthAccount> listaccount;
	// 数据库在手机内存中的目录
	private String DB_phone_path = "/data"
			+ Environment.getDataDirectory().getAbsolutePath()
			+ "/com.fenghuo.notes/databases";
	// 外部存储的备份目录
	private String DB_sd_path = Environment.getExternalStorageDirectory()
			.getPath() + "/Android/backups/com.fenghuo.notes";
	private String DB_name = "data.db";

	public BackupRestoreUtils(Context context) {
		this.context = context;
	}

	// 第一种思路 将所有数据转换成json字符串 存储到内存中

	public boolean backup() {
		alarmHelper = new DBAlarmHelper(context);
		kindHelper = new DbKindHelper(context);
		noteHelper = new DBNoteHelper(context);
		accountHelper = new DBAccountHelper(context);
		listalarm = alarmHelper.Getall();
		listKind = kindHelper.Getlist();
		listnote = noteHelper.Getlist();
		listaccount = accountHelper.GetListAll();
		// 转换json
		String Jsonalarms = "";
		if (listalarm != null && listalarm.size() > 0) {
			HashMap<String, String> map;
			for (Alarm alarm : listalarm) {
				map = new HashMap<String, String>();
				map.put("id", alarm.getId() + "");
				map.put("date", alarm.getId() + "");
				map.put("time", alarm.getId() + "");
				map.put("repeat", alarm.getId() + "");
				map.put("week", alarm.getId() + "");
				map.put("vibration", alarm.getId() + "");
				map.put("ringuri", alarm.getId() + "");
				map.put("content", alarm.getId() + "");
				JSONObject object = new JSONObject(map);
				Jsonalarms = Jsonalarms + object.toString() + "\n";
			}
		}

		return false;
	}

	public boolean restore() {

		return false;
	}

	public boolean checkexist() {

		File dri = new File(DB_sd_path);
		if (!dri.exists()) {
			dri.mkdirs();
			return false;
		}
		File file = new File(DB_sd_path + "/" + DB_name);
		if (file.exists()) {
			// 如果存在 ，读取数据信息
			// SQLiteDatabase.CursorFactory cursorFactory = new
			// SQLiteDatabase.CursorFactory() {
			// @Override
			// public Cursor newCursor(SQLiteDatabase db,
			// SQLiteCursorDriver masterQuery, String editTable,
			// SQLiteQuery query) {
			// return null;
			// }
			// };
			// SQLiteDatabase db = SQLiteDatabase.openDatabase(DB_sd_path,
			// cursorFactory, 0);

			return true;
		}
		return false;
	}

	// 第二种思路 直接将数据库导出来 -1 失败 0 未找到数据 1 成功

	public int restore2() {
		String pathto = DB_phone_path + "/" + DB_name;
		String pathfrom = DB_sd_path + "/" + DB_name;
		try {
			File to = new File(pathto);
			File from = new File(pathfrom);
			// 先决断sd卡文件夹是否存在
			File pderectroy = new File(DB_sd_path);
			if (!pderectroy.exists()) {
				pderectroy.mkdirs();
				return 0;
			}
			// sd卡数据库是否存在
			if (!from.exists()) {
				return 0;
			}
			FileInputStream streami = new FileInputStream(from);
			FileOutputStream streamo = new FileOutputStream(to);
			byte[] bs = new byte[512];
			int count = 0;
			while ((count = streami.read(bs)) > 0) {
				streamo.write(bs, 0, count);
			}
			streami.close();
			streamo.close();
			System.out.println("恢复成功===========");
			return 1;
		} catch (Exception e) {
			e.printStackTrace();
		}
		return -1;
	}

	public boolean backup2() {
		String pathto = DB_phone_path + "/" + DB_name;
		String pathfrom = DB_sd_path + "/" + DB_name;
		try {
			File from = new File(pathfrom);
			File to = new File(pathto);
			FileInputStream streami = new FileInputStream(to);
			FileOutputStream streamo = new FileOutputStream(from, false);
			byte[] bs = new byte[512];
			int count = 0;
			while ((count = streami.read(bs)) > 0) {
				streamo.write(bs, 0, count);
			}
			streami.close();
			streamo.close();
			System.out.println("备份成功！------------");
			return true;

		} catch (Exception e) {
			e.printStackTrace();
		}
		return false;
	}
}

File: src/com/fenghuo/notes/db/BackupRestoreUtils.kt
package com.fenghuo.notes.db

import android.content.Context
import android.os.Environment
import android.util.Log
import com.fenghuo.notes.bean.Alarm
import com.fenghuo.notes.bean.Kind
import com.fenghuo.notes.bean.MonthAccount
import com.fenghuo.notes.bean.Note
import com.fenghuo.notes.upload.CloudUtils
import com.fenghuo.utils.FileUtils
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.util.*

class BackupRestoreUtils(private val context: Context) {
    private val DB_SD_PATH = Environment.getExternalStorageDirectory().path + "/Android/backups/" + context.packageName
    private val DB_NAME = "data.db"

    companion object {
        const val SUCCESS = 0
        const val ERROR_FILE_NOT_FOUND = 1
        const val ERROR_UNKNOWN = 2
    }

    /***
     * 当前sdk卡是否有备份
     * */
    fun isExternalDBExist(): Boolean {
        return File(DB_SD_PATH + File.separator + DB_NAME).exists()
    }

    //本地备份
    fun backupLocal(): Boolean {
        val pathFrom = context.getDatabasePath(DB_NAME).absolutePath
        val pathTo = DB_SD_PATH + File.separator + DB_NAME

        Log.i(CloudUtils.TAG, "备份数据库: from=$pathFrom to=$pathTo")

        if (!File(pathFrom).exists()) {
            return false
        }

        return writeFile(pathFrom, pathTo)
    }

    //本地恢复
    fun restoreLocal(): Int {
        val pathTo = context.getDatabasePath(DB_NAME).absolutePath
        val pathFrom = DB_SD_PATH + File.separator + DB_NAME

        Log.i(CloudUtils.TAG, "恢复数据库: from=$pathFrom to=$pathTo")

        if (!File(pathFrom).exists()) {
            return ERROR_FILE_NOT_FOUND
        }
        context.databaseList().forEach {
            context.deleteDatabase(it)
            Log.d(CloudUtils.TAG, "刪除当前数据库:  " + context.getDatabasePath(it))
        }

        return if (writeFile(pathFrom, pathTo)) SUCCESS else ERROR_UNKNOWN
    }

    private fun writeFile(from: String, to: String): Boolean {
        try {
            val toFile = File(to)
            val fromFile = File(from)

            if (!fromFile.exists()) {
                return false
            }
            if (!toFile.exists()) {
                toFile.createNewFile()
            }
            val streamFrom = FileInputStream(from)
            val streamTo = FileOutputStream(to)
            val bs = ByteArray(512)
            var count: Int = streamFrom.read(bs)
            while (count > 0) {
                streamTo.write(bs, 0, count)
                count = streamFrom.read(bs)
            }
            streamFrom.close()
            streamTo.close()
            return true
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return false
    }

    private fun readFileEncrypt(from: String, to: String): Boolean {
        try {
            val toFile = File(to)
            val fromFile = File(from)

            if (!fromFile.exists()) {
                return false
            }
            if (!toFile.exists()) {
                toFile.createNewFile()
            }
            val streamFrom = FileInputStream(from)
            val streamTo = FileOutputStream(to)
            val head = ByteArray(128)
            val bs = ByteArray(512)
            //跳过head
            streamFrom.read(head)
            var count = streamFrom.read(bs)
            while (count > 0) {
                streamTo.write(bs, 0, count)
                count = streamFrom.read(bs)
            }
            streamFrom.close()
            streamTo.close()
            return true
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return false
    }

    private fun createFileEncrypt(from: String, to: String): Boolean {
        try {
            val fromFile = File(from)
            FileUtils.delete(to)
            val streamFrom = FileInputStream(fromFile)
            val streamTo = FileOutputStream(to, false)
            //add head
            val head = ByteArray(128)
            val random = Random(System.currentTimeMillis())
            random.nextBytes(head)
            streamTo.write(head)
            //add content
            val bs = ByteArray(512)
            var count = streamFrom.read(bs)
            while (count > 0) {
                streamTo.write(bs, 0, count)
                count = streamFrom.read(bs)
            }
            streamFrom.close()
            streamTo.close()
            return true

        } catch (e: Exception) {
            e.printStackTrace()
        }

        return false
    }

    //解密并恢复
    fun restoreEncryptFile(fromFile: String): Int {
        val pathTo = context.getDatabasePath(DB_NAME).absolutePath

        Log.d(CloudUtils.TAG, "恢复数据库: pathTo=$pathTo pathFrom=$fromFile")

        if (!File(fromFile).exists()) {
            return ERROR_FILE_NOT_FOUND
        }
        context.databaseList().forEach {
            context.deleteDatabase(it)
            Log.d(CloudUtils.TAG, "刪除当前数据库:  " + context.getDatabasePath(it))
        }

        return if (readFileEncrypt(fromFile, pathTo)) SUCCESS else ERROR_UNKNOWN
    }

    //防君子不防小人的加密
    fun copyDbFileWithEncrypt(toFileName: String): Boolean {
        val pathto = toFileName
        val pathfrom = context.getDatabasePath(DB_NAME)
        if (!pathfrom.exists()) {
            return false
        }
        val res = createFileEncrypt(pathfrom.absolutePath, pathto)
        Log.d(CloudUtils.TAG, "创建加密数据库: pathTo=$toFileName ")
        return res
    }
}
